viam-cartographer.go

```diff
@@ -127,63 +193,247 @@ func New(
+	if err != nil {
+		cartoSvc.logger.Errorw("cartofacade start failed", "error", err)
+		termErr := cf.Terminate(cancelCtx, cartoSvc.cartoFacadeTimeout)
+		if termErr != nil {
+			cartoSvc.logger.Errorw("cartofacade terminate failed", "error", termErr)
+			return termErr
+		}
+		return err
+	}
+	cartoSvc.cartofacade = &cf
+
+	return nil
+}
+
+// terminateCartoFacade
+// 1. stops & terminates carto facade
+// TODO: How does this work if we are closing down the context which manages the carto facade?
+func terminateCartoFacade(ctx context.Context, cartoSvc *cartographerService) error {
+	if cartoSvc.cartofacade == nil {
+		cartoSvc.logger.Debug("terminateCartoFacade called when cartoSvc.cartofacade is nil")
+		return nil
+	}
+	stopErr := cartoSvc.cartofacade.Stop(ctx, cartoSvc.cartoFacadeTimeout)
+	if stopErr != nil {
+		cartoSvc.logger.Errorw("cartofacade stop failed", "error", stopErr)
+	}
+
+	err := cartoSvc.cartofacade.Terminate(ctx, cartoSvc.cartoFacadeTimeout)
+	if err != nil {
+		cartoSvc.logger.Errorw("cartofacade terminate failed", "error", err)
+		return err
+	}
+	return stopErr
+}
+
+func initCartoGrpcServer(ctx, cancelCtx context.Context, cartoSvc *cartographerService) error {
 	if cartoSvc.primarySensorName != "" {
-		if err := dim2d.ValidateGetAndSaveData(cancelCtx, cartoSvc.dataDirectory, lidar,
-			sensorValidationMaxTimeoutSec, sensorValidationIntervalSec, cartoSvc.logger); err != nil {
-			return nil, errors.Wrap(err, "getting and saving data failed")
+		if err := dim2d.ValidateGetAndSaveData(cancelCtx, cartoSvc.dataDirectory, cartoSvc.lidar,
+			cartoSvc.sensorValidationMaxTimeoutSec, cartoSvc.sensorValidationIntervalSec, cartoSvc.logger); err != nil {
+			return errors.Wrap(err, "getting and saving data failed")
 		}
-		cartoSvc.StartDataProcess(cancelCtx, lidar, nil)
-		logger.Debugf("Reading data from sensor: %v", cartoSvc.primarySensorName)
+		cartoSvc.StartDataProcess(cancelCtx, cartoSvc.lidar, nil)
+		cartoSvc.logger.Debugf("Reading data from sensor: %v", cartoSvc.primarySensorName)
 	}
 
 	if err := cartoSvc.StartSLAMProcess(ctx); err != nil {
-		return nil, errors.Wrap(err, "error with slam service slam process")
+		return errors.Wrap(err, "error with slam service slam process")
 	}
 
-	client, clientClose, err := vcConfig.SetupGRPCConnection(ctx, cartoSvc.port, dialMaxTimeoutSec, logger)
+	client, clientClose, err := vcConfig.SetupGRPCConnection(ctx, cartoSvc.port, cartoSvc.dialMaxTimeoutSec, cartoSvc.logger)
 	if err != nil {
-		return nil, errors.Wrap(err, "error with initial grpc client to slam algorithm")
+		return errors.Wrap(err, "error with initial grpc client to slam algorithm")
 	}
 	cartoSvc.clientAlgo = client
 	cartoSvc.clientAlgoClose = clientClose
 
-	success = true
-	return cartoSvc, nil
+	return nil
 }
 
 // cartographerService is the structure of the slam service.
 type cartographerService struct {
 	resource.Named
 	resource.AlwaysRebuild
 	primarySensorName string
+	lidar             lidar.Lidar
```

Comment:
these attributes are added to the cartographerService as they are immutable & reduce the number of parameters that need to be passed throughout this package.